README.rst
.. _python-sdk:

**********
Python SDK
**********

General informations
====================

The Python library exposes all API methods described in :ref:`cloud-api` section.

You can download the python library on `GitHub <http://github.com/dailymotion/cloudkey-py>`_.

A master class named ``CloudKey`` exposes all objects (eg: :ref:`media-api`) through object attributes.

The library exposes all the remote methods of :ref:`cloud-api` and also have local methods for specific purpose like uploading files or getting signed stream URLs.

Remote methods
--------------

For instance, to call the ``list`` method from the ``media`` object, the code would be as follow

.. code-block:: python

  from cloudkey import CloudKey

  USER_ID = '4c1a4d3edede832bfd000002'
  API_KEY = '52a32c7890338770e3ea1601214c02142d297298'

  cloudkey = CloudKey(USER_ID, API_KEY, BASE_URL)
  cloudkey.media.list()

For methods expecting parameters, these must be passed as named arguments:

.. code-block:: python

   clouckey.media.list(fields=['id'], per_page=20, page=2)

The returned values are a direct mapping of the JSON structure into python's native types.

Local methods
-------------

File object
^^^^^^^^^^^

.. function:: upload_file(file)

   Upload a local file to Dailymotion Cloud using its path.

  :param file: the path of the file
  :type file: str
  :returns: A dict containing the ``url`` where the uploaded file can be accessed
  :rtype: dict

Media object
^^^^^^^^^^^^

.. function:: get_embed_url(id, seclevel=None, asnum=None, ip=None, useragent=None, expires=None)

   This method returns a signed URL to a Dailymotion Cloud player embed (see the API reference for details).

   The generated URL is perishable, and access is granted based on the provided security level bitmask.

   :param id: the id of the new media object.
   :type id: media ID
   :param seclevel: the security level bitmask (default is ``SecLevel.NONE``, see below for details).
   :type seclevel: int
   :param expires: the UNIX epoch expiration time (default is ``time() + 7200`` (2 hours from now)).
   :type expires: int

   The following arguments may be required if the ``SecLevel.DELEGATE`` option is not specified in the seclevel parameter, depending on the other options. This is not recommanded as it would probably lead to spurious access denials, mainly due to GeoIP databases discrepancies.

   :param asnum: the client's autonomous system number (default is ``None``).
   :type asnum: str
   :param ip: the client's IP adress (default is ``None``).
   :type ip: str
   :param useragent: the client's HTTP User-Agent header (default is ``None``).
   :type useragent: str

   Example::

      // Create an embed URL limited only to the AS of the end-user and valid for 1 hour
      url = cloudkey.media.get_embed_url(id=media['id'], seclevel=SecLevel.DELEGATE | SecLevel.ASNUM, expires=time() + 3600)
   
.. function:: get_stream_url(id, asset_name='mp4_h264_aac', seclevel=None, asnum=None, ip=None, useragent=None, expires=None)

   This method returns a signed URL to a Dailymotion Cloud video stream (see the API reference for details).

   The generated URL is perishable, and access is granted based on the provided security level bitmask.

   :param id: the id of the new media object.
   :type id: media ID
   :param asset_name: the desired media asset asset_name name (default is ``mp4_h264_aac``).
   :type asset_name: str
   :param seclevel: the security level bitmask (default is ``SecLevel.NONE``, see below for details).
   :type seclevel: int
   :param expires: the UNIX epoch expiration time (default is ``time() + 7200`` (2 hours from now)).
   :type expires: int

   The following arguments may be required if the ``SecLevel.DELEGATE`` option is not specified in the seclevel parameter, depending on the other options. This is not recommanded as it would probably lead to spurious access denials, mainly due to GeoIP databases discrepancies.

   :param asnum: the client's autonomous system number (default is ``None``).
   :type asnum: str
   :param ip: the client's IP adress (default is ``None``).
   :type ip: str
   :param useragent: the client's HTTP User-Agent header (default is ``None``).
   :type useragent: str


Quick Tour
==========


Security level options
======================

The security level defines the mechanism used by the Dailymotion Cloud architecture to ensure a mediastream
URL access will be limited to a single user or a group of users. The different (combinable) options are:

 - ``SecLevel.NONE``: the URL access is granted to everyone.
 - ``SecLevel.ASNUM``: the URL access is granted to the specified AS number only. AS numbers stands for 'Autonomous System number' and roughly map groups of IP to telcos and large organizations on the Internet (each ISP has its own AS number for instance, Dailyotion's AS number is AS41690).
 - ``SecLevel.IP``: the URL access is granted to the specified IP address only. This option may lead to spurious access denials as some users are load-balanced behind multiple proxies when accessing the Internet (this is mostly the case with ISPs and large organizations).
 - ``SecLevel.USERAGENT``: the URL access is granted to users sending the specified User-Agent HTTP header only.
 - ``SecLevel.DELEGATE``: the ASNUM, IP and User-Agent values are to be gathered at the server side during the first URL access and don't need to be specified at the client side beforehand (this is the recommanded approach as it will ensure a 100%-accurate ASNUM recognition).
 - ``SecLevel.USEONCE``: the URL access is granted once only (using this option will probably prevent seeking from working correctly).

For more information, please refer to the Dailymotion Cloud streams security documentation.

Exceptions
==========

 * RPCException: This is the base of all exceptions

   * TransportException: When an error occured with the HTTP transport
   * SerializerError: When the Request or the Response is not valid JSON
   * InvalidRequest: When the Request is not wellformed
   * InvalidCall: When the value of the ``call`` argument is invalid
   * InvalidObject: When you access an object that doesn't exist
   * InvalidMethod: When you access a method that doesn't exist
   * InvalidParameter: When a method is called with a invalid or missing parameter
   * AuthenticationError: When authentication information is invalid

     * RateLimitExceeded: When you exceed the number of API calls on a specific timeframe

   * ApplicationException: The base class of the following exceptions

     * NotFound: When action is requested on an item that doesn't exist
     * Exists: When action is requested on an item that already exists
     * LimitExceeded: When you reach the maximum number of allowed objects.


debian/changelog
cloudkey (1.2.7-1~karmic1) karmic-dc; urgency=low

  * New release

 -- Sebastien Estienne <sebastien.estienne@gmail.com>  Sat, 22 Sep 2012 00:45:51 +0200

cloudkey (1.2.6-1~karmic1) karmic-dc; urgency=low

  * New release

 -- Sebastien Estienne <sebastien.estienne@gmail.com>  Sun, 12 Aug 2012 21:44:01 +0200

cloudkey (1.2.4-1~karmic1) karmic-dc; urgency=low

  * Added protocol parameter to get_stream_url

 --  <jererc@gmail.com>  Thu, 12 Jan 2012 18:15:23 +0100

cloudkey (1.2.3-1~karmic1) karmic-dc; urgency=low

  * Added filename parameter to get_stream_url
  * Updated download url handling in get_stream_url
  * Added unit tests

 --  <jererc@gmail.com>  Wed, 04 Jan 2012 11:05:42 +0100

cloudkey (1.2.2-1~karmic1) karmic-dc; urgency=low

  * Added a version parameter for stream url

 --  <jererc@gmail.com>  Thu, 24 Nov 2011 14:51:56 +0100

cloudkey (1.2.1-1~karmic1) karmic-dc; urgency=low

  * cosmetic fix

 -- Sebastien Estienne <sebastien.estienne@gmail.com>  Sun, 17 Jul 2011 15:57:49 +0200

cloudkey (1.2.0-1~karmic1) karmic-dc; urgency=low

  * Support for live streaming

 -- Sebastien Estienne <sebastien.estienne@gmail.com>  Fri, 06 May 2011 20:27:59 +0200

cloudkey (1.1.4-1~karmic1) karmic-dc; urgency=low

  * get_stream_url returns static urls for thumbnail

 -- Sebastien Estienne <sebastien.estienne@gmail.com>  Wed, 30 Mar 2011 13:08:13 +0200

cloudkey (1.1.3~karmic1) karmic-dc; urgency=low

  * New release

 -- Sebastien Estienne <sebastien.estienne@gmail.com>  Mon, 28 Feb 2011 14:11:54 +0100

cloudkey (1.1~karmic1) karmic-dc; urgency=low

  * Cloud repository

 -- Sebastien Estienne <sebastien.estienne@gmail.com>  Thu, 24 Feb 2011 01:54:47 +0100

cloudkey (1.1-1) unstable; urgency=low

  * Support new API

 -- Sebastien Estienne <sebastien.estienne@dailymotion.com>  Sat, 25 Sep 2010 14:29:04 +0200

cloudkey (1.0-1) unstable; urgency=low

  * source package automatically created by stdeb 0.6.0+git

 -- unknown <unknown@unknown>  Fri, 10 Sep 2010 01:07:47 +0200


tests.py
try:
    import unittest2 as unittest
except ImportError, e:
    import unittest

from urllib import urlencode

from mock import patch, Mock

import cloudkey


def no_sign_url(*args, **kwargs):
    '''Mock the sign_url function, and return the url unmodified.
    '''
    return args[0]


class GetStreamUrl(unittest.TestCase):

    def setUp(self):
        self.client = Mock()
        self.client._user_id = '1' * 24
        self.client._api_key = '9' * 42

        self.media = cloudkey.MediaObject(self.client, 'name')
        self.id = '2' * 24
        self.cdn_url = 'http://cdn.dmcloud.net'
        self.static_url = 'http://static.dmcloud.net'

    def test_get_stream_url_static(self):
        asset_name = 'jpeg_thumbnail_medium'
        extension = 'jpeg'
        cdn_url = 'http://cdn.dmcloud.net'

        with patch.object(cloudkey, 'sign_url') as sign_url_mock:
            sign_url_mock.side_effect = no_sign_url
            res = self.media.get_stream_url(self.id, asset_name=asset_name, cdn_url=self.cdn_url)

        self.assertEquals(res, '%s/%s/%s/%s.%s' % (self.static_url, self.client._user_id, self.id, asset_name, extension))
        self.assertFalse(sign_url_mock.called)

    def test_get_stream_url_sign(self):
        asset_name = 'mp4_h264_aac'
        extension = 'mp4'

        with patch.object(cloudkey, 'sign_url') as sign_url_mock:
            sign_url_mock.side_effect = no_sign_url
            res = self.media.get_stream_url(self.id, asset_name=asset_name, cdn_url=self.cdn_url)

        self.assertEquals(res, '%s/route/%s/%s/%s.%s' % (self.cdn_url, self.client._user_id, self.id, asset_name, extension))
        self.assertEquals(sign_url_mock.call_args[0], (res, self.client._api_key))

    def test_get_stream_url_version(self):
        asset_name = 'mp4_h264_aac'
        extension = 'mp4'
        version = 123456

        with patch.object(cloudkey, 'sign_url') as sign_url_mock:
            sign_url_mock.side_effect = no_sign_url
            res = self.media.get_stream_url(self.id, asset_name=asset_name, version=version, cdn_url=self.cdn_url)

        self.assertEquals(res, '%s/route/%s/%s/%s-%s.%s' % (self.cdn_url, self.client._user_id, self.id, asset_name, version, extension))
        self.assertTrue(sign_url_mock.called)

    def test_get_stream_url_timestamp_not_static(self):
        asset_name = 'mp4_h264_aac'
        extension = 'mp4'
        expires = '123456'

        with patch.object(cloudkey, 'sign_url') as sign_url_mock:
            sign_url_mock.side_effect = no_sign_url
            res = self.media.get_stream_url(self.id, asset_name=asset_name, expires=expires, cdn_url=self.cdn_url)

        self.assertEquals(res, '%s/route/%s/%s/%s.%s' % (self.cdn_url, self.client._user_id, self.id, asset_name, extension))
        self.assertTrue(sign_url_mock.called)

    def test_get_stream_url_timestamp_static(self):
        asset_name = 'jpeg_thumbnail_medium'
        extension = 'jpeg'
        version = 123456

        with patch.object(cloudkey, 'sign_url') as sign_url_mock:
            sign_url_mock.side_effect = no_sign_url
            res = self.media.get_stream_url(self.id, asset_name=asset_name, version=version, cdn_url=self.cdn_url)

        self.assertEquals(res, '%s/%s/%s/%s-%s.%s' % (self.static_url, self.client._user_id, self.id, asset_name, version, extension))
        self.assertFalse(sign_url_mock.called)

    def test_get_stream_url_download_no_filename(self):
        asset_name = 'mp4_h264_aac'
        extension = 'mp4'
        download = True
        filename = None

        with patch.object(cloudkey, 'sign_url') as sign_url_mock:
            sign_url_mock.side_effect = no_sign_url
            res = self.media.get_stream_url(self.id, asset_name=asset_name, download=download, filename=filename, cdn_url=self.cdn_url)

        self.assertEquals(res, '%s/route/http/%s/%s/%s.%s' % (self.cdn_url, self.client._user_id, self.id, asset_name, extension))
        self.assertTrue(sign_url_mock.called)

    def test_get_stream_url_filename_no_download(self):
        asset_name = 'mp4_h264_aac'
        extension = 'mp4'
        download = False
        filename = 'test_&!filename.mp4'

        with patch.object(cloudkey, 'sign_url') as sign_url_mock:
            sign_url_mock.side_effect = no_sign_url
            res = self.media.get_stream_url(self.id, asset_name=asset_name, download=download, filename=filename, cdn_url=self.cdn_url)

        self.assertEquals(res, '%s/route/http/%s/%s/%s.%s?%s' % (self.cdn_url, self.client._user_id, self.id, asset_name, extension, urlencode({'filename': filename.encode('utf-8', 'ignore')})))
        self.assertTrue(sign_url_mock.called)

    def test_get_stream_url_bad_protocol(self):
        asset_name = 'mp4_h264_aac'
        extension = 'mp4'
        protocol = 'bad'

        with patch.object(cloudkey, 'sign_url') as sign_url_mock:
            sign_url_mock.side_effect = no_sign_url
            self.assertRaises(cloudkey.InvalidParameter, self.media.get_stream_url, self.id, asset_name=asset_name, protocol=protocol, cdn_url=self.cdn_url)

    def test_get_stream_url_protocol_http(self):
        asset_name = 'mp4_h264_aac'
        extension = 'mp4'
        protocol = 'http'

        with patch.object(cloudkey, 'sign_url') as sign_url_mock:
            sign_url_mock.side_effect = no_sign_url
            res = self.media.get_stream_url(self.id, asset_name=asset_name, protocol=protocol, cdn_url=self.cdn_url)

        self.assertEquals(res, '%s/route/%s/%s/%s/%s.%s' % (self.cdn_url, protocol, self.client._user_id, self.id, asset_name, extension))
        self.assertTrue(sign_url_mock.called)


class GetEmbedUrl(unittest.TestCase):

    def setUp(self):
        self.client = Mock()
        self.client._user_id = '1' * 24
        self.client._api_key = '9' * 42
        self.client._base_url = 'http://base.url'
        self.client._secure_base_url = 'https://base.url'

        self.media = cloudkey.MediaObject(self.client, 'name')
        self.id = '2' * 24

    def test_get_embed_url(self):
        with patch.object(cloudkey, 'sign_url') as sign_url_mock:
            sign_url_mock.side_effect = no_sign_url
            res = self.media.get_embed_url(self.id)

        self.assertEquals(res, '%s/embed/%s/%s' % (self.client._base_url, self.client._user_id, self.id))
        self.assertEquals(sign_url_mock.call_args[0], (res, self.client._api_key))

    def test_get_embed_url_https(self):
        with patch.object(cloudkey, 'sign_url') as sign_url_mock:
            sign_url_mock.side_effect = no_sign_url
            res = self.media.get_embed_url(self.id, secure=True)

        self.assertEquals(res, '%s/embed/%s/%s' % (self.client._secure_base_url, self.client._user_id, self.id))
        self.assertEquals(sign_url_mock.call_args[0], (res, self.client._api_key))


class GetQtrefUrl(unittest.TestCase):

    def setUp(self):
        self.client = Mock()
        self.client._user_id = '1' * 24
        self.client._api_key = '9' * 42
        self.client._base_url = 'http://base.url'

        self.media = cloudkey.MediaObject(self.client, 'name')
        self.id = '2' * 24

    def test_get_embed_url(self):
        with patch.object(cloudkey, 'sign_url') as sign_url_mock:
            sign_url_mock.side_effect = no_sign_url
            res = self.media.get_qtref_url(self.id)

        self.assertEquals(res, '%s/stream/%s/%s.mov' % (self.client._base_url, self.client._user_id, self.id))
        self.assertEquals(sign_url_mock.call_args[0], (res, self.client._api_key))

if __name__ == '__main__':
    unittest.main()


examples/upload_html/upload.py
BASE_URL='http://sebest.api.dev.int.dmcloud.net'
USER_ID='4cadc039dede832941000000'
API_KEY='f728debf25de13d367c2b73a6647979dd4f234e6'

from cloudkey import CloudKey
from cloudkey import AuthenticationError, NotFound, InvalidParameter, Exists, MissingParameter

from flask import Flask
from flask import render_template, request, url_for

app = Flask(__name__)

cloudkey = CloudKey(USER_ID, API_KEY, base_url=BASE_URL)

# The page with the upload FORM
@app.route("/")
def upload():
    # this is the URL that will recieve the information at the end of the upload
    target = request.base_url + url_for('upload_done')
    
    # this API call will return the upload url and the status url
    upload = cloudkey.file.upload(target=target, status=True, jsonp_cb='?')

    # this is the URL where we will upload our file
    upload_url = upload['url']

    # this is the URL that we poll using JSONP to get the upload status
    upload_status = upload['status']

    # the HTML page with the upload FORM
    return render_template('upload.html', upload_url=upload_url, upload_status=upload_status)

# The page called at the end of the upload
# request.args contrains the query string parameters
@app.route("/done")
def upload_done():
    # the url that you will use for media.create
    url = request.args['url']
    
    # the filename, you can use it as a title
    name = request.args['name']

    # we create a new media
    media = cloudkey.media.create(url=url, meta={'title': name})

    # the media id that you can store in your own DB
    media_id = media['id']
    return ''

if __name__ == "__main__":
    app.run(host='0.0.0.0', port=5003, debug=True)


cloudkey.py
"""Dailymotion Cloud RPC loosely based on JSON-RPC"""

import sys

API_ENDPOINT = '/api'

__version__ = "1.2.6"
__python_version__ = '.'.join([str(i) for i in sys.version_info[:3]])
_DEBUG = False

import os
import StringIO
import string
import random
import hashlib
import datetime
import time
import pycurl
import zlib
import re
import base64
import urllib
try:
    import json
except ImportError:
    import simplejson as json

#
# Common
#
class SecLevel:
    NONE      = 0
    DELEGATE  = 1 << 0
    ASNUM     = 1 << 1
    IP        = 1 << 2
    USERAGENT = 1 << 3
    USEONCE   = 1 << 4
    COUNTRY   = 1 << 5
    REFERER   = 1 << 6

def sign_url(url, secret, seclevel=None, asnum=None, ip=None, useragent=None, countries=None, referers=None, expires=None):
    # Normalize parameters
    seclevel = seclevel or SecLevel.NONE
    expires = int(expires or time.time() + 7200)

    # Compute digest
    (url, unused, query) = url.partition('?')
    secparams = ''
    public_secparams = []
    if not seclevel & SecLevel.DELEGATE:
        if seclevel & SecLevel.ASNUM:
            if not asnum:
                raise ValueError('ASNUM security level required and no AS number provided.')
            secparams += asnum
        if seclevel & SecLevel.IP:
            if not ip:
                raise ValueError('IP security level required and no IP address provided.')
            secparams += ip
        if seclevel & SecLevel.USERAGENT:
            if not useragent:
                raise ValueError('USERAGENT security level required and no user-agent provided.')
            secparams += useragent
        if seclevel & SecLevel.COUNTRY:
            if not countries or len(countries) == 0:
                raise ValueError('COUNTRY security level required and no coutry list provided.')
            if type(countries) is not list:
                raise ValueError('Invalid format for COUNTRY, should be a list of country codes.')
            if countries[0] == '-':
                countries = '-' + ','.join(countries[1:])
            else:
                countries = ','.join(countries)
            if not re.match(r'^-?(?:[a-zA-Z]{2})(?:,[a-zA-Z]{2})*$', countries):
                raise ValueError('Invalid format for COUNTRY security level parameter.')
            public_secparams.append('cc=%s' % countries.lower())
        if seclevel & SecLevel.REFERER:
            if not referers or len(referers) == 0:
                raise ValueError('REFERER security level required and no referer list provided.')
            if type(referers) is not list:
                raise ValueError('Invalid format for REFERER, should be a list of url strings.')
            public_secparams.append('rf=%s' % urllib.quote_plus(' '.join([referer.replace(' ', '%20') for referer in referers])))

    public_secparams_encoded = ''
    if len(public_secparams) > 0:
        public_secparams_encoded = base64.urlsafe_b64encode(zlib.compress('&'.join(public_secparams)))
    rand = ''.join(random.choice(string.ascii_lowercase + string.digits) for unused in range(8))
    digest = hashlib.md5('%d%s%d%s%s%s%s' % (seclevel, url, expires, rand, secret, secparams, public_secparams_encoded)).hexdigest()

    # Return signed URL
    return '%s?%sauth=%s-%s-%s-%s%s' % (url, (query + '&' if query else ''), expires, seclevel, rand, digest, ('-' + public_secparams_encoded if public_secparams_encoded else ''))

def normalize(arg=None):
    """Normalizes an argument for signing purpose.

    This is used for normalizing the arguments of RPC method calls.

    :param arg: The argument to normalize

    :return: A string representating the normalized argument.

    .. doctest::

     >>> from cloud.rpc import normalize
     >>> normalize(['foo', 42, 'bar'])
     'foo42bar'
     >>> normalize({'yellow': 1, 'red': 2, 'pink' : 3})
     'pink3red2yellow1'
     >>> normalize(['foo', 42, {'yellow': 1, 'red': 2, 'pink' : 3}, 'bar'])
     'foo42pink3red2yellow1bar'
     >>> normalize(None)
     ''
     >>> normalize([None, 1,2])
     '12'
     >>> normalize({2: [None, 1,2], 3: None, 4:5})
     '212345'
    """
    res = ''

    t_arg = type(arg)
    if t_arg in (list, tuple):
        for i in arg:
            res += normalize(i)
    elif t_arg is dict:
        keys = arg.keys()
        keys.sort()
        for key in keys:
            res += '%s%s' % (normalize(key), normalize(arg[key]))
    elif t_arg is unicode:
        res = arg.encode('utf8')
    elif t_arg is bool:
        res = 'true' if arg else 'false'
    elif arg != None:
        res = str(arg)

    return res

def sign(shared_secret, msg):
    """Signs a message using a shared secret.

    :param shared_secret: The shared secret used to sign the message
    :param msg: The message to sign

    :return: The signature as a string

    .. doctest::

     >>> from cloud.rpc import sign
     >>> sign('sEcReT_KeY', 'hello world')
     '5f048ebaf6f06576b60716dc8f815d85'
    """
    if isinstance(msg, unicode):
        msg = msg.encode('utf8')
    return hashlib.md5(msg + shared_secret).hexdigest()

DRM_TOKEN_TIMEOUT = 600

def gen_drm_token(user_id, media_id, api_key, rights=None, meta=None, callback_url=None, expires=0, max_replay=0, encode=True):
    info = {
        'user_id': user_id,
        'media_id': media_id,
        'expires': expires if expires and isinstance(expires, int) else int(time.time() + DRM_TOKEN_TIMEOUT),
        'nonce': base64.urlsafe_b64encode(os.urandom(8)),
    }
    if max_replay and isinstance(max_replay, int):
        info['max_replay'] = max_replay
    if isinstance(rights, dict):
        info['rights'] = rights
    if isinstance(meta, dict):
        info['meta'] = meta
    if isinstance(callback_url, (str, unicode)):
        info['callback_url'] = callback_url

    info['auth'] = sign(api_key, normalize(info))

    payload = json.dumps(info)
    if encode:
        return base64.urlsafe_b64encode(payload)
    return payload

#################################################################################
class RPCException(Exception):
    """Base class for all RPC exceptions.
    """
    code = 100

    def __init__(self, message, data=None):
        self.message = message
        self.data = data

    def __str__(self):
        msg = '%d: %s' % (self.code, self.message)
        if self.data:
            msg = '%s [%s]' % (msg, self.data)
        return msg

class ProcessorException(RPCException):
    """Base class for all Processor exceptions
    """
    code = 200

class TransportException(RPCException):
    """Exceptions in transport layer
    """
    code = 300

class AuthenticationError(RPCException):
    """Base class for all Auth exceptions.
    """
    code = 400

class RateLimitExceeded(AuthenticationError):
    code = 410

class SerializerError(RPCException):
    code = 500

class InvalidRequest(RPCException):
    code = 600

class InvalidObject(InvalidRequest):
    code = 610

class InvalidMethod(InvalidRequest):
    code = 620

class InvalidParameter(InvalidRequest):
    code = 630

class InvalidCall(InvalidRequest):
    code = 640

class MissingParameter(InvalidRequest):
    code = 650

class ApplicationException(RPCException):
    code = 1000

class NotFound(ApplicationException):
    code = 1010

class Exists(ApplicationException):
    code = 1020

class LimitExceeded(ApplicationException):
    code = 1030

def RPCException_handler(error):
    # TODO Autogenerate this dict by introspection
    exceptions = {
        200: ProcessorException,
        300: TransportException,
        400: AuthenticationError,
        410: RateLimitExceeded,
        500: SerializerError,
        600: InvalidRequest,
        610: InvalidObject,
        620: InvalidMethod,
        630: InvalidParameter,
        640: InvalidCall,
        650: MissingParameter,
        1000: ApplicationException,
        1010: NotFound,
        1020: Exists,
        1030: LimitExceeded,
        }

    e = exceptions.get(error['code'])
    if not e:
        e = RPCException
        e.code = error['code']
    return e(error['message'], error.get('data', None))


class JSONEncoder(json.JSONEncoder):
    """Extends JSON encoder to handle types like datetime

    .. doctest::

     >>> import json
     >>> from datetime import datetime
     >>> from cloud.rpc import JSONEncoder
     >>> my_date = datetime(year=2010, month=04, day=11)
     >>> json.dumps(my_date, cls=JSONEncoder)
     '1270936800'
    """
    def default(self, obj):
        if isinstance(obj, (datetime.datetime, datetime.date)):
            return int(time.mktime(obj.timetuple()))
        return json.JSONEncoder.default(self, obj)


class dotdict(dict):
    def __getattr__(self, attr):
        return self.get(attr, None)

class ClientObject(object):

    def __init__(self, client, name):
        self._client = client
        self._name = name

    def __getattr__(self, method):
        global _DEBUG

        def func(**kwargs):
            request = {
                'call': '%s.%s' % (self._name, method),
                'args': kwargs,
                }

            if not self._client._act_as_user:
                user_infos = self._client._user_id
            else:
                user_infos = "%s/%s" % (self._client._user_id, self._client._act_as_user)
            request['auth'] = user_infos + ':' + sign(self._client._api_key, user_infos + normalize(request))


            c = pycurl.Curl()
            c.setopt(pycurl.URL, str(self._client._api_endpoint))
            c.setopt(pycurl.USERAGENT, 'cloudkey-py/%s (Python %s)' % (__version__, __python_version__))
            c.setopt(pycurl.HTTPHEADER, ["Content-Type: application/json", 'Expect:'])

            if _DEBUG:
                print '   Example request::'
                print ''
                print '   Normalization or request: ' + normalize(request)
                print ''
                for line in  json.dumps(request, cls=JSONEncoder, indent=2).split('\n'):
                    print '       %s' % line
                print ''

            try:
                data = json.dumps(request, cls=JSONEncoder)
            except (TypeError, ValueError), e:
                raise SerializerError(str(e))

            c.setopt(pycurl.POSTFIELDS, data)

            if self._client._proxy:
                c.setopt(pycurl.PROXY, self._client._proxy)

            response = StringIO.StringIO()
            c.setopt(pycurl.WRITEFUNCTION, response.write)

            try:
                c.perform()
            except pycurl.error, e:
                raise TransportException(str(e))
            finally:
                c.close()

            try:
                #msg = json.loads(response.getvalue(), object_hook=lambda x: dotdict(x))
                msg = json.loads(response.getvalue())
                if _DEBUG:
                    print '   Example response::'
                    print ''
                    for line in  json.dumps(msg, indent=2).split('\n'):
                        print '       %s' % line
                    print ''
            except (TypeError, ValueError), e:
                raise SerializerError(str(e))
            error = msg.get('error', None)
            if error:
                raise RPCException_handler(error)
            return msg.get('result')

        return func


class FileObject(ClientObject):

    def upload_file(self, file, progress = None):
        if not os.path.exists(file):
            raise IOError("[Errno 2] No such file or directory: '%s'" % file)
        result = self.upload()

        if isinstance(file, unicode):
            file = file.encode('utf8')

        c = pycurl.Curl()
        c.setopt(pycurl.URL, str(result['url']))
        c.setopt(pycurl.USERAGENT, 'cloudkey-py/%s (Python %s)' % (__version__, __python_version__))
        c.setopt(pycurl.HTTPHEADER, ['Expect:'])
        c.setopt(pycurl.FOLLOWLOCATION, True)
        c.setopt(pycurl.HTTPPOST, [('file', (pycurl.FORM_FILE, file))])

        if self._client._proxy:
            c.setopt(pycurl.PROXY, self._client._proxy)

        if progress:
            c.setopt(pycurl.NOPROGRESS, 0)
            c.setopt(pycurl.PROGRESSFUNCTION, lambda x, y, total, current: progress(current, total))

        response = StringIO.StringIO()
        c.setopt(pycurl.WRITEFUNCTION, response.write)

        try:
            c.perform()
        except pycurl.error, e:
            raise TransportException(str(e))
        c.close()

        return json.loads(response.getvalue())


class MediaObject(ClientObject):

    def _get_url(self, base_path, id, seclevel=None, asnum=None, ip=None, useragent=None, countries=None, referers=None, expires=None, skin=None, drm_token=None, secure=False):
        if type(id) not in (str, unicode):
            raise InvalidParameter('id is not valid')
        base_url = self._client._secure_base_url if secure else self._client._base_url
        url = '%s%s/%s/%s' % (base_url, base_path, self._client._user_id, id)
        url = sign_url(url, self._client._api_key, seclevel=seclevel, asnum=asnum, ip=ip, useragent=useragent, countries=countries, referers=referers, expires=expires) \
            + ('&skin=%s' % skin if skin else '')
        if drm_token:
            drm_token = gen_drm_token(self._client._user_id, id, self._client._api_key,
                                      expires=drm_token.get('expires'),
                                      rights=drm_token.get('rights'),
                                      meta=drm_token.get('meta'),
                                      max_replay=drm_token.get('max_replay'),
                                      callback_url=drm_token.get('callback_url'))
            url += '&drm_token=%s' % drm_token
        return url

    def get_embed_url(self, id, seclevel=None, asnum=None, ip=None, useragent=None, countries=None, referers=None, expires=None, skin=None, drm_token=None, secure=False):
        return self._get_url('/embed', id, seclevel, asnum, ip, useragent, countries, referers, expires, skin, drm_token, secure)

    def get_swf_url(self, id, seclevel=None, asnum=None, ip=None, useragent=None, countries=None, referers=None, expires=None, skin=None, drm_token=None, secure=False):
        return self._get_url('/player/swf', id, seclevel, asnum, ip, useragent, countries, referers, expires, skin, drm_token, secure)

    def get_qtref_url(self, id, seclevel=None, asnum=None, ip=None, useragent=None, countries=None, referers=None, expires=None):
        if type(id) not in (str, unicode):
            raise InvalidParameter('id is not valid')
        url = '%s/stream/%s/%s.mov' % (self._client._base_url, self._client._user_id, id)
        return sign_url(url, self._client._api_key, seclevel=seclevel, asnum=asnum, ip=ip, useragent=useragent, countries=countries, referers=referers, expires=expires)

    def get_stream_url(self, id, asset_name='mp4_h264_aac', seclevel=None, asnum=None, ip=None, useragent=None, countries=None, referers=None, expires=None, download=False, filename=None, version=None, protocol=None, cdn_url='http://cdn.dmcloud.net'):
        if type(id) not in (str, unicode):
            raise InvalidParameter('id is not valid')
        if protocol not in (None, 'hls', 'rtmp', 'hps', 'http', 'ss', 'hds'):
            raise InvalidParameter('%s is not a valid streaming protocol' % protocol)
        version = '-%d' % version if version else ''
        if asset_name.startswith('jpeg_thumbnail_'):
            base_url = cdn_url.replace('cdn.', 'static.')
            return '%s/%s/%s/%s%s.jpeg' % (base_url, self._client._user_id, id, asset_name, version)
        extension = '.%s' % asset_name.split('_')[0]
        if download or filename:
            protocol = 'http'

        if asset_name == 'abs':
            extension = ''
            if not protocol:
                raise InvalidParameter('protocol is required for abs asset_name')

        url = '%s/route%s/%s/%s/%s%s%s' % (cdn_url, '/%s' % protocol if protocol else '', self._client._user_id, id, asset_name, version, extension)

        if filename:
            url = '%s?%s' % (url, urllib.urlencode({'filename': filename.encode('utf-8', 'ignore')}))

        return sign_url(url, self._client._api_key, seclevel=seclevel, asnum=asnum, ip=ip, useragent=useragent, countries=countries, referers=referers, expires=expires)


class CloudKey(object):

    def __init__(self, user_id, api_key, base_url='http://api.dmcloud.net', proxy=None, debug=False):
        global _DEBUG
        self._user_id = user_id if user_id else ''
        self._api_key = api_key if api_key else ''
        self._base_url = base_url
        self._secure_base_url = base_url.replace('http:', 'https:')
        self._act_as_user = None
        self._api_endpoint =  base_url + API_ENDPOINT
        self._proxy = proxy
        _DEBUG = debug

    def __getattr__(self, method):
        if method == 'file':
            return FileObject(self, method)
        if method == 'media':
            return MediaObject(self, method)
        return ClientObject(self, method)

    def act_as_user(self, user):
        self._act_as_user = user
